File: apps/mobile/android/app/src/main/java/dev/kneecap/app/stt/WhisperTranscriber.java
package dev.kneecap.app.stt;

import android.content.Context;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;

import org.json.JSONException;

/**
 * kneecap M10 — the native-plugin-facing entry point for
 * `NativeBridge.transcribe()`. Produces exactly the wire shape
 * `packages/native-bridge/src/capacitor-bridge.ts` expects
 * (`NativeTranscribeResult` there): a list of segments, each carrying its
 * own start/end/text plus a `tokens` array shaped like `RawWordTiming`
 * (`caption-smoothing.ts`) — `coarseStartMicros`/`coarseEndMicros` from
 * whisper.cpp's `t0`/`t1`, `dtwStartMicros` from `t_dtw` (or `null` for its
 * `-1` sentinel), `confidence` from the token's decode probability. The
 * mandatory smoothing pass itself runs entirely on the TS side
 * (`mapNativeTranscribeResult()` in `capacitor-bridge.ts`) — this class's
 * only job is producing honest raw numbers for it to smooth.
 *
 * HONEST STATUS (see also `WhisperJNI`'s class doc comment): this class is
 * real, reviewed Java, but two things stand between it and an actual
 * on-device transcription today:
 *
 *   1. No `libkneecap_whisper.so` is built/bundled — `WhisperJNI`'s native
 *      calls throw `UnsatisfiedLinkError`.
 *   2. `decodeToMono16kFloat()` below is NOT implemented — turning an
 *      arbitrary native-custody audio/video URI into 16kHz mono float32
 *      PCM (whisper.cpp's hard input requirement) needs a real
 *      MediaExtractor/MediaCodec decode+resample pipeline. That is
 *      legitimately its own unit of work, and overlaps with M4's media
 *      pipeline (which will already own audio-track extraction for proxy
 *      generation) — implementing it twice, differently, here would be
 *      wasted work. Throws `UnsupportedOperationException` naming this
 *      reason rather than silently returning empty/fake audio.
 *
 * `transcribe()` therefore always throws today. That is the correct,
 * honest state for M10 part 1 — see the plan's own explicit allowance for
 * "simulator/device runs may be partial" and this class's own tests (none
 * yet: exercising this requires either a real `.so` + real audio, or a
 * device/emulator run, neither available in the session that wrote this).
 */
public final class WhisperTranscriber {

	private WhisperTranscriber() {}

	public static final class NotYetWiredException extends RuntimeException {
		public NotYetWiredException(String message) {
			super(message);
		}
	}

	/**
	 * @param context     Android context, used to resolve the bundled model
	 *                    asset path (`assets/models/ggml-{modelSize}.en.bin`,
	 *                    populated at build time by
	 *                    `scripts/download-whisper-model.sh` — see that
	 *                    script and `assets/models/README.md`).
	 * @param audioUri    native-custody handle to the source clip's audio
	 *                    (matches `MediaHandle.uri` — never a `blob:` URL).
	 * @param modelSize   "tiny" or "base" (`TranscribeOptions.modelSize`);
	 *                    resolves to the bundled `.en` asset — kneecap v1
	 *                    ships English-only models (see
	 *                    `download-whisper-model.sh`'s header comment).
	 * @param languageHint reserved for a future non-English model; unused
	 *                    while only `.en` models are bundled.
	 */
	public static JSObject transcribe(
			Context context, String audioUri, String modelSize, String languageHint)
			throws JSONException {
		String modelAssetPath = "models/ggml-" + modelSize + ".en.bin";
		if (!modelAssetExists(context, modelAssetPath)) {
			throw new NotYetWiredException(
					"Bundled model asset '"
							+ modelAssetPath
							+ "' not found. Run scripts/download-whisper-model.sh at build time"
							+ " (see apps/mobile/android/app/src/main/assets/models/README.md).");
		}

		// The two real gaps this class's doc comment describes. Both throw
		// before reaching WhisperJNI, so a missing .so is never the FIRST
		// error a caller sees — the audio-decode gap is more informative
		// and is fixed first regardless of library-loading status.
		float[] pcm16kMono = decodeToMono16kFloat(context, audioUri);

		WhisperJNI.ensureLibraryLoaded();
		// WHISPER_AHEADS_TINY_EN = 3, WHISPER_AHEADS_BASE_EN = 5 — ordinals
		// of `enum whisper_alignment_heads_preset` in `include/whisper.h`,
		// read directly off that header while writing this (see
		// WhisperJNI.initContext's doc comment for why an int ordinal
		// crosses the JNI boundary instead of redeclaring the enum here).
		int aheadsPreset = "base".equals(modelSize) ? 5 : 3;
		long ctx =
				WhisperJNI.initContext(
						resolveAbsoluteModelPath(context, modelAssetPath), /* dtw= */ true, aheadsPreset);
		if (ctx == 0) {
			throw new IllegalStateException("whisper_init failed for model " + modelAssetPath);
		}
		try {
			WhisperJNI.fullTranscribe(ctx, preferredThreadCount(), pcm16kMono);
			return buildResult(ctx);
		} finally {
			WhisperJNI.freeContext(ctx);
		}
	}

	private static JSObject buildResult(long ctx) throws JSONException {
		JSObject result = new JSObject();
		JSArray segments = new JSArray();
		int segmentCount = WhisperJNI.getSegmentCount(ctx);
		for (int s = 0; s < segmentCount; s++) {
			JSObject segment = new JSObject();
			segment.put("startMicros", WhisperJNI.getSegmentT0(ctx, s) * 10_000L);
			segment.put("endMicros", WhisperJNI.getSegmentT1(ctx, s) * 10_000L);
			segment.put("text", WhisperJNI.getSegmentText(ctx, s));
			segment.put("confidence", JSObject.NULL);

			JSArray tokens = new JSArray();
			int tokenCount = WhisperJNI.getSegmentTokenCount(ctx, s);
			for (int t = 0; t < tokenCount; t++) {
				JSObject token = new JSObject();
				token.put("text", WhisperJNI.getTokenText(ctx, s, t));
				token.put("coarseStartMicros", WhisperJNI.getTokenT0(ctx, s, t) * 10_000L);
				token.put("coarseEndMicros", WhisperJNI.getTokenT1(ctx, s, t) * 10_000L);
				long dtw = WhisperJNI.getTokenDtw(ctx, s, t);
				token.put("dtwStartMicros", dtw == -1 ? JSObject.NULL : dtw * 10_000L);
				token.put("confidence", WhisperJNI.getTokenProbability(ctx, s, t));
				tokens.put(token);
			}
			segment.put("tokens", tokens);
			segments.put(segment);
		}
		result.put("segments", segments);
		return result;
	}

	private static boolean modelAssetExists(Context context, String assetPath) {
		try (java.io.InputStream unused = context.getAssets().open(assetPath)) {
			return true;
		} catch (java.io.IOException e) {
			return false;
		}
	}

	private static String resolveAbsoluteModelPath(Context context, String assetPath) {
		// whisper.cpp's `whisper_init_from_file` needs a real filesystem
		// path, not an APK asset stream — the actual implementation copies
		// the asset to `context.getFilesDir()` once (cached across calls)
		// before calling into WhisperJNI. Not implemented in this session;
		// see class doc comment — `transcribe()` never reaches this call in
		// practice today because `decodeToMono16kFloat` throws first.
		throw new NotYetWiredException(
				"resolveAbsoluteModelPath: asset-to-filesystem-cache copy not implemented yet (M10 follow-up).");
	}

	/**
	 * See class doc comment, gap #2. Deliberately throws with a specific,
	 * actionable message instead of returning a zero-filled or truncated
	 * buffer — a caller silently getting back "silence" would be a much
	 * worse failure mode than a clear, immediate exception.
	 */
	private static float[] decodeToMono16kFloat(Context context, String audioUri) {
		throw new NotYetWiredException(
				"Audio decode to 16kHz mono float32 PCM is not implemented yet (M10 follow-up;"
						+ " overlaps with M4's media pipeline — see WhisperTranscriber's class doc"
						+ " comment). audioUri="
						+ audioUri);
	}

	private static int preferredThreadCount() {
		// Matches examples/whisper.android's own WhisperCpuConfig heuristic
		// (leave one core free for the OS/UI thread), without importing
		// that reference project — this is a two-line equivalent, not
		// worth a dependency.
		int cores = Runtime.getRuntime().availableProcessors();
		return Math.max(1, cores - 1);
	}
}

File: apps/mobile/ios/App/App/WhisperTranscriber.swift
import Foundation

/// kneecap M10 — the iOS half of `NativeBridge.transcribe()`, plan M10
/// items 1-2 ("whisper.cpp as the cross-platform engine, native plugin...
/// iOS: build with -DWHISPER_COREML=1"). Deliberately parallel in shape to
/// `apps/mobile/android/app/src/main/java/dev/kneecap/app/stt/WhisperTranscriber.java`
/// — same wire contract, same two honestly-open gaps, same reasoning for
/// why each throws instead of faking success.
///
/// STATUS — READ BEFORE WIRING THIS INTO THE APP TARGET:
///
/// This file is real, reviewed Swift, written directly against the real
/// whisper.h C API (verified locally: `brew install whisper-cpp` gave
/// whisper.cpp 1.9.2 with a real `/opt/homebrew/include/whisper.h` this
/// session cross-checked every struct field and function signature
/// against). It is INTENTIONALLY NOT added to `App.xcodeproj/project.pbxproj`
/// — this Xcode project uses explicit `PBXFileReference`/`PBXBuildFile`
/// entries (not Xcode 16's synchronized-folder groups), so a `.swift` file
/// dropped into this directory is inert until explicitly wired in. That is
/// deliberate: adding `import whisper` to a file the build graph DOES
/// compile (e.g. editing the already-wired `NativeBridgePlugin.swift`
/// directly) before `whisper.xcframework` is actually embedded would break
/// the M3 CI build (`.github/workflows/mobile-ci.yml`'s
/// `ios-simulator-build` job) for every other track working in this repo
/// right now. Wiring this in for real requires, in order:
///
///   1. Run whisper.cpp's own `./build-xcframework.sh` (from a whisper.cpp
///      checkout — NOT part of this repo, referenced the same way the
///      plan's Core ML step references `generate-coreml-model.sh`) to
///      produce `whisper.xcframework`.
///   2. Embed that xcframework in `App.xcodeproj` (drag into the project,
///      "Embed & Sign" on the App target — the same one-real-manual-step
///      pattern as plan M10 item 2's Core ML `.mlmodelc` signing).
///   3. Add THIS file to the App target's Sources build phase (again, one
///      manual Xcode step, or a careful hand-edit of `project.pbxproj`
///      adding matching `PBXFileReference`/`PBXBuildFile`/sources-phase
///      entries — not attempted this session; hand-editing `.pbxproj` by
///      script is a real, separate risk class, the same reason M13's
///      signing work stays in CI rather than an agent session).
///   4. Register a `transcribe` `CAPPluginMethod` on
///      `NativeBridgePlugin.swift` that calls into this file — NOT done
///      yet either, for the same reason as step 3: that file IS currently
///      compiled by CI and stays exactly as M3 left it until steps 1-3 are
///      real.
///
/// Two further gaps, same as Android, kept honest rather than faked:
///
///   - `decodeToMono16kFloat` is not implemented (AVFoundation
///     `AVAssetReader` + a sample-rate/channel-count converter — real
///     work, overlaps with M4's media pipeline exactly like the Android
///     side).
///   - Runtime iOS-26 detection for `SpeechAnalyzer`/`SpeechTranscriber`
///     (plan M10 item 2's second half) is not implemented in this file at
///     all — it belongs in `NativeBridgePlugin.swift` as a preferred path
///     ahead of this whisper.cpp fallback, once this file is wired in.
enum WhisperTranscriberError: Error {
	case modelAssetNotFound(String)
	case notYetWired(String)
	case contextInitFailed(String)
}

enum WhisperTranscriber {

	/// - Parameters:
	///   - audioUri: native-custody handle to the source clip's audio
	///     (matches `MediaHandle.uri` — never a `blob:` URL).
	///   - modelSize: "tiny" or "base" (`TranscribeOptions.modelSize`);
	///     resolves to the bundled `.en` asset in
	///     `App/App/Resources/models/ggml-{modelSize}.en.bin` — see that
	///     directory's README and `scripts/download-whisper-model.sh`.
	/// - Returns: the same wire shape
	///   `packages/native-bridge/src/capacitor-bridge.ts`'s
	///   `NativeTranscribeResult` expects — a dictionary ready for
	///   `CAPPluginCall.resolve(_:)`, once a `transcribe` plugin method
	///   exists to call this (see this file's header comment, step 4).
	static func transcribe(
		audioUri: String,
		modelSize: String,
		languageHint: String?
	) throws -> [String: Any] {
		guard let modelURL = bundledModelURL(modelSize: modelSize) else {
			throw WhisperTranscriberError.modelAssetNotFound(
				"ggml-\(modelSize).en.bin not found in App/App/Resources/models — "
					+ "run scripts/download-whisper-model.sh before building."
			)
		}

		// Same ordering rationale as the Android side: the audio-decode gap
		// is more informative than "whisper module not linked," so it's
		// checked first even though, structurally, both must be fixed
		// before this ever returns real data.
		_ = try decodeToMono16kFloat(audioUri: audioUri)

		throw WhisperTranscriberError.notYetWired(
			"WhisperTranscriber.transcribe: this file is not yet part of the App target's "
				+ "build (see this file's header comment) — reaching this line at all would "
				+ "mean the xcframework, pbxproj wiring, and audio decode are ALL done, which "
				+ "they are not as of M10 part 1. modelURL=\(modelURL.path)"
		)

		// The shape a real implementation returns, once wired — kept here
		// as a comment (not dead code the compiler would need to reconcile
		// with the `throw` above) so the wire contract stays documented
		// next to the one place that produces it:
		//
		// return [
		//     "segments": segments.map { segment in
		//         [
		//             "startMicros": segment.t0 * 10_000,
		//             "endMicros": segment.t1 * 10_000,
		//             "text": segment.text,
		//             "confidence": NSNull(),
		//             "tokens": segment.tokens.map { token in
		//                 [
		//                     "text": token.text,
		//                     "coarseStartMicros": token.t0 * 10_000,
		//                     "coarseEndMicros": token.t1 * 10_000,
		//                     "dtwStartMicros": token.t_dtw == -1 ? NSNull() : token.t_dtw * 10_000,
		//                     "confidence": token.p,
		//                 ]
		//             },
		//         ]
		//     }
		// ]
	}

	private static func bundledModelURL(modelSize: String) -> URL? {
		Bundle.main.url(
			forResource: "ggml-\(modelSize).en",
			withExtension: "bin",
			subdirectory: "models"
		)
	}

	/// See this file's header comment. Throws with a specific, actionable
	/// reason rather than returning empty/fake PCM.
	private static func decodeToMono16kFloat(audioUri: String) throws -> [Float] {
		throw WhisperTranscriberError.notYetWired(
			"Audio decode to 16kHz mono float32 PCM is not implemented yet (M10 follow-up; "
				+ "AVAssetReader + resampling, overlaps with M4's media pipeline). audioUri=\(audioUri)"
		)
	}
}
